### PR TITLE
Mobile JS refactor

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: app/assets/javascripts/mobile/mobile_drawer.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        evt.preventDefault();
+        $("#app").toggleClass("draw");
+      });
+
+      this.followedTags.find("a").bind("tap click", function(evt){
+        evt.preventDefault();
+        self.followedTags.find("li").toggleClass("hide");
+      });
+    }
+  }
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_drawer.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    like: function(likeCounter, link){
+      $.ajax({
+        url: href,
+        dataType: "json",
+        type: "POST",
+        beforeSend: this.showLoader(link),
+        success: function(data){
+          this.removeLoader(link);
+          link.attr("href", href + "/" + data["id"]);
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: ['id'] is better written in dot notation.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            self.removeLoader(link);
+          },
+          error: function(){
+            self.removeLoader(link);
+            alert(Diaspora.I18n.t("failed_to_reshare"));
+          }
+        });
+      }
+    }
+  }
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    removeLoader: function(link){
+      link.removeClass("loading")
+          .toggleClass("active")
+          .toggleClass("inactive");
+    },
+
+    like: function(likeCounter, link){
+      $.ajax({
+        url: href,
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: 'href' is not defined.

File: app/assets/javascripts/mobile/mobile_post_actions.js (hunk 2 above)
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: 'href' is not defined.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+          if(likeCounter){
+            likeCounter.text(parseInt(likeCounter.text) + 1);
+          }
+        }
+      });
+    },
+
+    unlike: function(likeCounter, link){
+      $.ajax({
+        url: href,
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: 'href' is not defined.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        dataType: "json",
+        type: "DELETE",
+        beforeSend: this.showLoader(link),
+        success: function(){
+          this.removeLoader(link);
+          link.attr("href", href.replace(/\/\d+$/, ""));
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: 'href' is not defined.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            likeCounter.text(parseInt(likeCounter.text) - 1);
+          }
+        }
+      });
+    },
+
+    onLike: function(evt){
+      evt.preventDefault();
+      var link = $(evt.target),
+          href = link.attr("href"),
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: 'href' is defined but never used.

**Comment by AugierLe42e** [comment]
Hmm? No, no, no, it's not orphan at all. It just overlaps #6509 so I'm waiting for it to be merged to fix this one ;)

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+      link.addClass("loading");
+    },
+
+    hideLoader: function(link){
+      link.removeClass("loading");
+    },
+
+    toggleActive: function(link) {
+      link.toggleClass("active")
+          .toggleClass("inactive")
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+          likeCounter.text(parseInt(likeCounter.text(), 10) + 1);
+        }
+      };
+
+      $.ajax({
+        url: url,
+        dataType: "json",
+        type: "POST",
+        beforeSend: function() {
+          Diaspora.Mobile.PostActions.showLoader(link)
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        },
+        success: onSuccess,
+        complete: function() {
+          Diaspora.Mobile.PostActions.hideLoader(link)
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+          likeCounter.text(parseInt(likeCounter.text(), 10) - 1);
+        }
+      };
+
+      $.ajax({
+        url: url,
+        dataType: "json",
+        type: "DELETE",
+        beforeSend: function() {
+          Diaspora.Mobile.PostActions.showLoader(link)
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+        },
+        success: onSuccess,
+        complete: function() {
+          Diaspora.Mobile.PostActions.hideLoader(link)
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+          href = link.attr("href"),
+          confirmText = link.attr("title");
+
+      if(!link.hasClass("loading") && link.hasClass("inactive") && confirm(confirmText)) {
+        $.ajax({
+          url: href + "&provider_display_name=mobile",
+          dataType: "json",
+          type: "POST",
+          beforeSend: function() {
+            Diaspora.Mobile.PostActions.showLoader(link)
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+          },
+          success: function() {
+            Diaspora.Mobile.PostActions.toggleActive(link)
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+          },
+          error: function() {
+            alert(Diaspora.I18n.t("failed_to_reshare"))
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

File: app/assets/javascripts/mobile/mobile_post_actions.js [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+          },
+          complete: function() {
+            Diaspora.Mobile.PostActions.hideLoader(link)
```
Comment (diaspora-code-review) [inline app/assets/javascripts/mobile/mobile_post_actions.js]: Missing semicolon.

**Comment by AugierLe42e** [comment]
Iz ok for review.

**Comment by svbergerem** [comment]
I didn't do a detailed review yet but the code is looking quite good. Please make sure that you didn't miss anything in your refactoring. At least the `autosize` call is missing. I'd also like to see some tests like I introduced them in https://github.com/diaspora/diaspora/commit/3bc5b673c6f34122583814279956d25c10c9d2b2 while refactoring the mobile comments.

**Comment by AugierLe42e** [comment]
I planned to write some tests just in the next PR but I can do it in this one. I'll do a careful check tonight.

**Comment by AugierLe42e** [comment]
@svbergerem: [The new tests](https://github.com/diaspora/diaspora/pull/6394/files#diff-3a9e7b970ccd89fc1388badfe82219b7R1) I've written [are failing](https://travis-ci.org/diaspora/diaspora/jobs/108491222#L712). It looks like the fixtures are not correctly loaded so HTML is empty. Am I missing something?

**Comment by svbergerem** [comment]
@AugierLe42e the fixtures are loading fine. You can test this by adding `expect($("#drawer").length).toBeGreaterThan(0)` after loading the fixtures and then running `bin/rake jasmine:ci` locally.

The problem is that you define `allAspects` etc in the drawer code before running the initialize function which is called after the page loaded and the fixtures have been added to the page. You can fix this by moving the definition into the initialize function:
```diff
diff --git a/app/assets/javascripts/mobile/mobile_drawer.js b/app/assets/javascripts/mobile/mobile_drawer.js
index 547d703..746776a 100644
--- a/app/assets/javascripts/mobile/mobile_drawer.js
+++ b/app/assets/javascripts/mobile/mobile_drawer.js
@@ -1,10 +1,10 @@
 (function(){
   Diaspora.Mobile.Drawer = {
-    allAspects: $("#all_aspects"),
-    followedTags: $("#followed_tags"),
-    menuBadge: $("#menu-badge"),
-
     initialize: function(){
+      this.allAspects = $("#all_aspects");
+      this.followedTags = $("#followed_tags");
+      this.menuBadge = $("#menu-badge");
+
       this.allAspects.bind("tap click", function(evt){
         evt.preventDefault();
         $(this).find("+ li").toggleClass("hide");
```

**Comment by AugierLe42e** [comment]
It's ok here. Ready for another review.

File: app/assets/javascripts/mobile/mobile_drawer.js [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+  Diaspora.Mobile.Drawer = {
+    allAspects: function(){
+      return $("#all_aspects")
+    },
+    followedTags: function(){
+      return $("#followed_tags")
+    },
+    menuBadge: function(){
+      return $("#menu-badge")
+    },
```
Comment (svbergerem) [inline app/assets/javascripts/mobile/mobile_drawer.js]: Why do you use those functions?

Comment (AugierLe42e) [inline app/assets/javascripts/mobile/mobile_drawer.js]: There was probably a reason, at some point. Now I guess I can directly use the jQuery in the `initialize`.

File: spec/javascripts/mobile/mobile_application_spec.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+    it("calls autosize for textareas", function(){
+      Diaspora.Mobile.initialize();
+      expect(window.autosize).toHaveBeenCalled();
+      expect(window.autosize.calls.mostRecent().args[0].selector).toBe("textarea");
+    });
+
+    it("deactivates shield", function(){
+      Diaspora.Mobile.initialize();
+      var $shield = $(".stream_element").first();
```
Comment (svbergerem) [inline spec/javascripts/mobile/mobile_application_spec.js]: Please check that this has the `shield-active` class so we know that it really changes something.

File: spec/javascripts/mobile/mobile_post_actions_spec.js [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+      this.link = $(".stream .like-action").first();
+    });
+
+    it("adds and removes loading class", function(){
+      expect(this.link).not.toHaveClass("loading");
+      Diaspora.Mobile.PostActions.showLoader(this.link);
+      expect(this.link).toHaveClass("loading");
+      Diaspora.Mobile.PostActions.hideLoader(this.link);
+      expect(this.link).not.toHaveClass("loading");
+    });
```
Comment (svbergerem) [inline spec/javascripts/mobile/mobile_post_actions_spec.js]: this tests the `showLoader` and `hideLoader` functions so I'd move that into a `describe("showLoader and hideLoader", function(){});` block.

File: spec/javascripts/mobile/mobile_post_actions_spec.js [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+
+    it("doesn't do anything if link is loading", function(){
+      Diaspora.Mobile.PostActions.showLoader(this.link);
+      expect(this.link).toHaveClass("loading");
```
Comment (svbergerem) [inline spec/javascripts/mobile/mobile_post_actions_spec.js]: Since we already test above that `showLoader` does the right thing, you could also add the `loading` class to the link manually. Then this test would also work if the `showLoader` function breaks.

File: spec/javascripts/mobile/mobile_post_actions_spec.js [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+
+    it("calls like if like button is inactive", function(){
+      expect(this.link).toHaveClass("inactive");
+      this.link.click();
+      expect(Diaspora.Mobile.PostActions.like).toHaveBeenCalled();
+    });
+
+    it("calls unlike if like button is active", function(){
+      Diaspora.Mobile.PostActions.toggleActive(this.link);
+      expect(this.link).toHaveClass("active");
```
Comment (svbergerem) [inline spec/javascripts/mobile/mobile_post_actions_spec.js]: Again, just toggle the classes manually so we only test the `onLike` function.

File: spec/javascripts/mobile/mobile_post_actions_spec.js [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+
+      Diaspora.Mobile.PostActions.like(this.likeCounter, this.link);
+      jasmine.Ajax.requests.mostRecent().respondWith({status: 400});
+      expect(Diaspora.Mobile.PostActions.hideLoader).toHaveBeenCalled();
+      Diaspora.Mobile.PostActions.like(this.likeCounter, this.link);
+      jasmine.Ajax.requests.mostRecent().respondWith({status: 201, responseText: "{\"id\": \"1\"}"});
+      expect(Diaspora.Mobile.PostActions.hideLoader).toHaveBeenCalled();
+    });
+
+    it("doesn't do anything on error", function(){
```
Comment (svbergerem) [inline spec/javascripts/mobile/mobile_post_actions_spec.js]: > doesn't do anything

is not correct. Maybe

> doesn't activate the link on error

?

File: spec/javascripts/mobile/mobile_post_actions_spec.js [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+      expect(Diaspora.Mobile.PostActions.toggleActive).not.toHaveBeenCalled();
+      expect(this.likeCounter.text()).toBe("0");
+    });
+
+    it("activates link on success", function(){
+      spyOn(Diaspora.Mobile.PostActions, "toggleActive");
+      var data = this.link.data("url");
+
+      Diaspora.Mobile.PostActions.like(this.likeCounter, this.link);
+      jasmine.Ajax.requests.mostRecent().respondWith({status: 201, responseText: "{\"id\": \"1\"}"});
```
Comment (svbergerem) [inline spec/javascripts/mobile/mobile_post_actions_spec.js]: Maybe use another id which is different from the like counter.

File: spec/javascripts/mobile/mobile_post_actions_spec.js [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+
+      Diaspora.Mobile.PostActions.unlike(this.likeCounter, this.link);
+      jasmine.Ajax.requests.mostRecent().respondWith({status: 400});
+      expect(Diaspora.Mobile.PostActions.hideLoader).toHaveBeenCalled();
+      Diaspora.Mobile.PostActions.unlike(this.likeCounter, this.link);
+      jasmine.Ajax.requests.mostRecent().respondWith({status: 204});
+      expect(Diaspora.Mobile.PostActions.hideLoader).toHaveBeenCalled();
+    });
+
+    it("doesn't do anything on error", function(){
```
Comment (svbergerem) [inline spec/javascripts/mobile/mobile_post_actions_spec.js]: Again I'd rephrase this.

File: spec/javascripts/mobile/mobile_post_actions_spec.js [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+      jasmine.Ajax.requests.mostRecent().respondWith({status: 201, responseText: "{}"});
+      expect(Diaspora.Mobile.PostActions.toggleActive).toHaveBeenCalledWith(this.reshareLink);
+    });
+
+    it("pops an alert on error", function(){
+      this.reshareLink.click();
+      jasmine.Ajax.requests.mostRecent().respondWith({status: 400});
+      expect(window.alert).toHaveBeenCalledWith(Diaspora.I18n.t("failed_to_reshare"));
+    });
+  })
```
Comment (svbergerem) [inline spec/javascripts/mobile/mobile_post_actions_spec.js]: missing semicolon

**Comment by AugierLe42e** [comment]
@svbergerem: I'm done.

**Comment by svbergerem** [comment]
Thanks a lot! The code is looking much better now.

**Comment by Flaburgan** [comment]
Thank you!

**Comment by AugierLe42e** [comment]
Thx guys.
